### DISCUSSION TITLE
Use of randomForestSRC package on GPU

### DISCUSSION THREAD

**Discussion by maxel62** [kogalur/randomForestSRC#230] [Q&A] [answered]
Hello,

I develop random forest algorithm on GPUs and I currently use _ranger_ package. I would like to use OTFI for missing values and the package _randomForestSRC_ seems to do this with _na.action = "na.impute"_ option. I read the package is optimized for CPUs with OpenMP. I would like to know if it is optimized for GPUs too ? 

Maxime

**Comment by kogalur** [comment, marked as answer]
randomForestSRC supports OTFI in a robust fashion, with na.action = "na.impute".  We have no plans for GPU support at this time.  However, OpenMP should work out of the box on all platforms.

**Reply by maxel62** [reply]
Thank you for your quick answer. 
So, I understand that it can work and be efficient on GPU, but it is managed by OpenMP. Is it right ? If OpenML is installed on GPUs, I can follow these instructions [https://www.randomforestsrc.org/articles/installation.html]() to install the package _randomForestSRC_ ?

**Reply by kogalur** [reply]
I don't understand what you are asking.  OpenMP is an API that implements shared memory multiprocessing among CPU cores.  Code has to be specifically written to use the OpenMP API and runs on a specific type of architecture -- namely a node with multiple CPU cores with shared memory.  That is what we have done with randomForestSRC. From what I understand, OpenML is a high level API that allows some form of interoperability among data sets, but it certainly doesn't get you into the realm of OpenMP, or GPGPU processing using CUDA or even OpenCL combining GPUs and CPUs.  All of these options require a very specific software design using those API's and must be compiled appropriately using CUDA or OpenCL capable compilers.  You will not be able to use GPGPU processing with randomForestSRC.
